blockchain/indexbuilder.go

```diff
@@ -113,12 +119,145 @@ func (ib *IndexBuilder) HandleBlock(blk *block.Block) error {
 	ib.pendingBlks <- blk
 	return nil
 }
+func initIndexActionsKey(store db.KVStore) error {
+	_, err := store.Get(blockActionBlockMappingNS, indexActionsTipIndexKey)
+	if err != nil && errors.Cause(err) == db.ErrNotExist {
+		if err = store.Put(blockActionBlockMappingNS, indexActionsTipIndexKey, make([]byte, 8)); err != nil {
+			return errors.Wrap(err, "failed to write initial value for tip index of index actions")
+		}
+	}
 
+	_, err = store.Get(blockActionBlockMappingNS, indexActionsTipHeightKey)
+	if err != nil && errors.Cause(err) == db.ErrNotExist {
+		if err = store.Put(blockActionBlockMappingNS, indexActionsTipHeightKey, make([]byte, 8)); err != nil {
+			return errors.Wrap(err, "failed to write initial value for tip height of index actions")
+		}
+	}
+	return err
+}
+func (ib *IndexBuilder) getStartHeightAndIndex() (startHeight, startIndex uint64, err error) {
+	// get index that already builded
+	startIndex, err = getNextIndex(ib.store)
+	if err != nil {
+		return
+	}
+	// get height that already builded
+	startHeight, err = getNextHeight(ib.store)
+	if err != nil {
+		return
+	}
+	return
+}
+func (ib *IndexBuilder) commitBatchAndClear(tipIndex, tipHeight uint64, batch db.KVStoreBatch) error {
+	tipIndexBytes := byteutil.Uint64ToBytes(tipIndex)
+	batch.Put(blockActionBlockMappingNS, indexActionsTipIndexKey, tipIndexBytes, "failed to put tip index of actions")
+	tipHeightBytes := byteutil.Uint64ToBytes(tipHeight)
+	batch.Put(blockActionBlockMappingNS, indexActionsTipHeightKey, tipHeightBytes, "failed to put tip height")
+	if err := ib.store.Commit(batch); err != nil {
+		return err
+	}
+	return nil
+}
+func (ib *IndexBuilder) initAndLoadActions() error {
+	err := initIndexActionsKey(ib.store)
+	if err != nil {
+		return err
+	}
+	tipHeight, err := ib.dao.getBlockchainHeight()
+	if err != nil {
+		return err
+	}
+	startHeight, startIndex, err := ib.getStartHeightAndIndex()
+	if err != nil {
+		return err
+	}
+	zap.L().Info("Loading actions", zap.Uint64("startHeight", startHeight), zap.Uint64("startIndex", startIndex))
+	batch := db.NewBatch()
+	i := startHeight
+	for ; i <= tipHeight; i++ {
+		hash, err := ib.dao.getBlockHash(i)
+		if err != nil {
+			return err
+		}
+		body, err := ib.dao.body(hash)
+		if err != nil {
+			return err
+		}
+		blk := &block.Block{
+			Body: *body,
+		}
+		err = indexBlockHash(startIndex, hash, ib.store, blk, batch)
+		if err != nil {
+			return err
+		}
+		receipts, err := ib.dao.getReceipts(i)
+		// if receipts are not available,this error will be ignored
+		if err != nil && errors.Cause(err) != db.ErrNotExist {
+			return err
+		}
+		putReceipts(i, receipts, batch)
+		startIndex += uint64(len(blk.Actions))
+		// commit once every 10000 heights
+		if i%10000 == 0 {
```

Comment:
```suggestion
		if i%10000 == 0 || i == tipHeight {
```

Then remove L211-L214